src/Grpc.AspNetCore.Web/Internal/Base64PipeReader.cs

```diff
@@ -0,0 +1,193 @@
+                        innerResult.IsCompleted);
+                }
+
+                // Attempt to get more data
+                _inner.AdvanceTo(innerResult.Buffer.Start, innerResult.Buffer.End);
+                innerResult = await _inner.ReadAsync(cancellationToken);
+            }
+
+            // Limit result to complete base64 segments (multiples of 4)
+            var buffer = innerResult.Buffer.Slice(0, (innerResult.Buffer.Length / 4) * 4);
+
+            // The content can contain multiple fragments of base64 content
+            // Check for padding, and limit returned data to one fragment at a time
+            var paddingIndex = PositionOf(buffer, (byte)'=');
+            if (paddingIndex != null)
+            {
+                _currentInnerBuffer = buffer.Slice(0, ((paddingIndex.Value / 4) + 1) * 4);
+            }
+            else
+            {
+                _currentInnerBuffer = buffer;
+            }
+
+            // TODO - optimize this
+            // Get array from a pool, or attempt to reuse array inside readonly sequence
+            var data = _currentInnerBuffer.ToArray();
+            var validLength = (data.Length / 4) * 4;
+            var status = Base64.DecodeFromUtf8InPlace(data.AsSpan(0, validLength), out var bytesWritten);
+            if (status == OperationStatus.Done || status == OperationStatus.NeedMoreData)
+            {
+                _currentDecodedBuffer = new ReadOnlySequence<byte>(data, 0, bytesWritten);
+                return new ReadResult(
+                    _currentDecodedBuffer,
+                    innerResult.IsCanceled,
+                    innerResult.IsCompleted);
+            }
+
+            throw new InvalidOperationException("Unexpected status: " + status);
+        }
+
+        public override bool TryRead(out ReadResult result)
+        {
+            throw new NotImplementedException();
+        }
+
+        private static int? PositionOf<T>(in ReadOnlySequence<T> source, T value) where T : IEquatable<T>
+        {
+            if (source.IsSingleSegment)
+            {
+                int index = source.First.Span.IndexOf(value);
+                if (index != -1)
+                {
+                    return index;
+                }
+
+                return null;
+            }
+            else
+            {
+                return PositionOfMultiSegment(source, value);
+            }
+        }
+
+        private static int? PositionOfMultiSegment<T>(in ReadOnlySequence<T> source, T value) where T : IEquatable<T>
+        {
+            SequencePosition position = source.Start;
+            int total = 0;
+            while (source.TryGet(ref position, out ReadOnlyMemory<T> memory))
+            {
+                int index = memory.Span.IndexOf(value);
+                if (index != -1)
+                {
+                    return total + index;
+                }
+                else if (position.GetObject() == null)
```

Comment:
Nit: Is this really necessary? Wouldn't the next TryGet fail in this case?

Reply:
It breaks out of the while loop so there won't be a next TryGet. This method duplicates what ReadOnlySequence does.

Reply:
Are you copying this from ReadOnlySequence.Enumerator.TryMove()? If so, I'm pretty sure the `_next.GetObject() == null` check is only there to protect against you already being at the end of the sequence. So if you need this check at all, you'd want to check if `source.Start.GetObject() == null` at the start of the method, not in the middle of the while loop when TryGet just returned true.

Reply:
It's the same as this - https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Memory/src/System/Buffers/BuffersExtensions.cs#L18-L56 - expect it returns an index relative to the start of the sequence instead of a position.